src/robot_state_publisher.cpp

```diff
@@ -136,11 +136,16 @@ RobotStatePublisher::RobotStatePublisher(const rclcpp::NodeOptions & options)
 
   setupURDF(urdf_xml);
 
+  auto subscriber_options = rclcpp::SubscriptionOptions();
+  subscriber_options.qos_overriding_options =
+    rclcpp::QosOverridingOptions::with_default_policies();
+
   // subscribe to joint state
   joint_state_sub_ = this->create_subscription<sensor_msgs::msg::JointState>(
-    "joint_states", 10, std::bind(
-      &RobotStatePublisher::callbackJointState, this,
-      std::placeholders::_1));
+    "joint_states",
+    rclcpp::SensorDataQoS(),
```

Comment:
I think I'd rather keep this as `reliable` by default, but you now have the option to switch it to `best-effort` if you want.  At least, that's the most compatible with the current setup and least surprising.

Reply:
I think `best-effort` makes sense as a default, and I wouldn't consider it a breaking change -- according to the table in the [QoS compatibilities](https://docs.ros.org/en/galactic/Concepts/About-Quality-of-Service-Settings.html#qos-compatibilities) section, a `best-effort` subscriber can still receive messages from a `reliable` publisher. Changing the default here lays the groundwork for allowing publishers to change their default in a non-breaking way in a future version.

Reply:
> according to the table in the [QoS compatibilities](https://docs.ros.org/en/galactic/Concepts/About-Quality-of-Service-Settings.html#qos-compatibilities) section, a `best-effort` subscriber can still receive messages from a `reliable` publisher.

While that is true, best-effort doesn't intuitively seem like the correct default for the `joint_state`.  If you miss a `joint_state` update, then you won't update the tf2 tree appropriately, and there is nothing that guarantees that the publisher will republish that joint in the near future.  That is, this doesn't seem to act like a Sensor, where you know you will get another update about the state of the world soon.

Reply:
The systems I've worked with publish `joint_state` at regular intervals, such as the [create_driver](https://github.com/AutonomyLab/create_robot/blob/f0a8dea560f83d148223f62bf8b0734775e09be3/create_driver/src/create_driver.cpp#L286) for my hobby Roomba and the platforms I've used at work.

That said, it sounds like you have some counterexamples in mind, where changing this default would cause issues. Even though I still think that `best-effort` would be a more suitable default in most situations, I'm happy to switch the default back in order to avoid such issues.